### PR TITLE
[Feature-2585][server] Support ssh command 

### PR DESCRIPTION
**Pull request by peilongwu**
### What is the purpose of the pull request
Support execute command on remote server. Only support linux.

### Brief change log
1. Add remote server datasource support.
2. Update shell task to support execute command on remote datasource

### CODE REVIEW

**Comment by yangyichao-mango** [comment]
Hi,

Please modify the pr title and add the pr content according the specifications[1]. Thx a lot for your good job~
[1] https://dolphinscheduler.apache.org/zh-cn/docs/development/pull-request.html

**Comment by peilongwu** [comment]
OK, Got it


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月23日(星期四) 下午3:20
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] Merge support ssh command (#3283)





 
Hi,
 
Please modify the pr title and add the pr content according the specifications[1]. Thx a lot for your good job~
 [1] https://dolphinscheduler.apache.org/zh-cn/docs/development/pull-request.html

**Comment by peilongwu** [comment]
------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月23日(星期四) 下午3:20
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] Merge support ssh command (#3283)





 
Hi,
 
Please modify the pr title and add the pr content according the specifications[1]. Thx a lot for your good job~
 [1] https://dolphinscheduler.apache.org/zh-cn/docs/development/pull-request.html

**Comment by peilongwu** [comment]
I already modify the pr title and comment.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月23日(星期四) 下午3:20
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] Merge support ssh command (#3283)





 
Hi,
 
Please modify the pr title and add the pr content according the specifications[1]. Thx a lot for your good job~
 [1] https://dolphinscheduler.apache.org/zh-cn/docs/development/pull-request.html

File: dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/DataSourceService2Test.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import java.util.Map;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = ApiApplicationServer.class)
+public class DataSourceService2Test {
```
Comment (yangyichao-mango) [inline dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/DataSourceService2Test.java:38]: Hi,
Why introduce this test case without adding this test to root pom?

**Comment by yangyichao-mango** [comment]
Hi,
We can not include and use `jsch` officially in DolphinScheduler because of the licenses.
image
image

**Comment by yangyichao-mango** [comment]
Hi,
Please make sure the 0.1.42 version license of `jsch` to avoid the license problem.
image
image

**Comment by yangyichao-mango** [comment]
Hi,
I switch to your branch and check the feature.
1. Remote machine resources may not be suitable for being put with dataSources. And we can talk about the remote machine resource manage place. [1]

If you have any question or suggestion, welcome to put forward.

image

**Comment by peilongwu** [comment]
I add this test in root pom before.&nbsp;


I forgot to add this test when I solved the conflicts in this time.



------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月23日(星期四) 晚上7:55
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283)





 
@yangyichao-mango commented on this pull request.
 
 
In dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/DataSourceService2Test.java:
 &gt; +import org.apache.dolphinscheduler.common.enums.UserType; +import org.apache.dolphinscheduler.dao.entity.User; +import org.junit.Assert; +import org.junit.Test; +import org.junit.runner.RunWith; +import org.slf4j.Logger; +import org.slf4j.LoggerFactory; +import org.springframework.beans.factory.annotation.Autowired; +import org.springframework.boot.test.context.SpringBootTest; +import org.springframework.test.context.junit4.SpringRunner; + +import java.util.Map; + +@RunWith(SpringRunner.class) +@SpringBootTest(classes = ApiApplicationServer.class) +public class DataSourceService2Test {  
Hi,
 Why introduce this test case without adding this test to root pom?

**Comment by peilongwu** [comment]
I do not import jsch in DolphinScheduler. When I use jsch, it's already in that.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月23日(星期四) 晚上10:42
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283)





 
Hi,
 We can not include and use jsch officially in DolphinScheduler because of the licenses.

**Comment by peilongwu** [comment]
I do not import jsch in DolphinScheduler. When I use jsch, it's already in that.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月23日(星期四) 晚上10:50
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283)





 
Hi,
 Please make sure the 0.1.42 version license of jsch to avoid the license problem.

**Comment by yangyichao-mango** [comment]
Hi,
I have checked the license of `jsch`, it is suitable for DolphinScheduler~




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月24日(星期五) 上午10:02
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"Yichao Yang"<1048262223@qq.com&gt;;"Mention"<mention@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283)





 I do not import jsch in DolphinScheduler. When I use jsch, it's already in that. 
 
 
 ------------------&amp;nbsp;原始邮件&amp;nbsp;------------------ 
 发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&amp;gt;; 
 发送时间:&amp;nbsp;2020年7月23日(星期四) 晚上10:50 
 收件人:&amp;nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&amp;gt;; 
 抄送:&amp;nbsp;"裴龙武"<peilongwu@qq.com&amp;gt;;"Author"<author@noreply.github.com&amp;gt;; 
 主题:&amp;nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283) 
 
 
 
 
 
   
 Hi, 
  Please make sure the 0.1.42 version license of jsch to avoid the license problem.

**Comment by peilongwu** [comment]
I do not have a good idea to add many remote servers. If we can not add these remote servers to data source, we must develop a new feature to manage remote server.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月23日(星期四) 晚上11:09
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283)





 
Hi,
 I switch to your branch and check the feature.
  
Remote machine resources may not be suitable for being put with dataSources. And we can talk about the remote machine resource manage place. [1]
  
If you have any question or suggestion, welcome to put forward.

**Comment by peilongwu** [comment]
OK, it's very&nbsp; good.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月24日(星期五) 上午10:04
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283)





 Hi, 
 I have checked the license of `jsch`, it is suitable for DolphinScheduler~ 
 
 
 
 
 ------------------&amp;nbsp;原始邮件&amp;nbsp;------------------ 
 发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&amp;gt;; 
 发送时间:&amp;nbsp;2020年7月24日(星期五) 上午10:02 
 收件人:&amp;nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&amp;gt;; 
 抄送:&amp;nbsp;"Yichao Yang"<1048262223@qq.com&amp;gt;;"Mention"<mention@noreply.github.com&amp;gt;; 
 主题:&amp;nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283) 
 
 
 
 
 
  I do not import jsch in DolphinScheduler. When I use jsch, it's already in that.  
   
   
  ------------------&amp;amp;nbsp;原始邮件&amp;amp;nbsp;------------------  
  发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&amp;amp;gt;;  
  发送时间:&amp;amp;nbsp;2020年7月23日(星期四) 晚上10:50  
  收件人:&amp;amp;nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&amp;amp;gt;;  
  抄送:&amp;amp;nbsp;"裴龙武"<peilongwu@qq.com&amp;amp;gt;;"Author"<author@noreply.github.com&amp;amp;gt;;  
  主题:&amp;amp;nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283)  
   
   
   
   
   
     
  Hi,  
   Please make sure the 0.1.42 version license of jsch to avoid the license problem.

**Comment by peilongwu** [comment]
It has a lot of cases to use many remote serves to execute some shell commands. So I think that we must to design a good method to support this feature.


I want to add a new table to manage the remote servers， but many committers feel that this change is too big. They think we can manage remote servers like data sources. So I develop this feature like this.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/incubator-dolphinscheduler"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年7月23日(星期四) 晚上11:09
收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;;
抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command  (#3283)





 
Hi,
 I switch to your branch and check the feature.
  
Remote machine resources may not be suitable for being put with dataSources. And we can talk about the remote machine resource manage place. [1]
  
If you have any question or suggestion, welcome to put forward.

**Comment by yangyichao-mango** [comment]
> It has a lot of cases to use many remote serves to execute some shell commands. So I think that we must to design a good method to support this feature. I want to add a new table to manage the remote servers， but many committers feel that this change is too big. They think we can manage remote servers like data sources. So I develop this feature like this.
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "apache/incubator-dolphinscheduler" <notifications@github.com&gt;; 发送时间:&nbsp;2020年7月23日(星期四) 晚上11:09 收件人:&nbsp;"apache/incubator-dolphinscheduler"<incubator-dolphinscheduler@noreply.github.com&gt;; 抄送:&nbsp;"裴龙武"<peilongwu@qq.com&gt;;"Author"<author@noreply.github.com&gt;; 主题:&nbsp;Re: [apache/incubator-dolphinscheduler] [Feature-2585][server] Support ssh command (#3283) Hi, I switch to your branch and check the feature. Remote machine resources may not be suitable for being put with dataSources. And we can talk about the remote machine resource manage place. [1] If you have any question or suggestion, welcome to put forward. — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub, or unsubscribe.

I got it.

**Comment by yangyichao-mango** [comment]
When I set the password free login, and I do not set the password in the remote server settings, the following error will appear. Do we have to set the password when we want to login in the remote server? In a lot of scene,  we already set the password free login. Can you add the password free login to the sshtask?

image
image

**Comment by dailidong** [comment]
@peilongwu , supporting ssh task will be an important feature, and could you add password free to ssh task?  by the way, please solve the conflicting files, thx

> Hi,
> When I set the password free login, and I do not set the password in the remote server settings, the following error will appear. Do we have to set the password when we want to login in the remote server? In a lot of scene, we already set the password free login. Can you add the password free login to the sshtask?
> 
> image
> image
> image

**Comment by peilongwu** [comment]
> @peilongwu , supporting ssh task will be an important feature, and could you add password free to ssh task? by the way, please solve the conflicting files, thx
> 
> > Hi,
> > When I set the password free login, and I do not set the password in the remote server settings, the following error will appear. Do we have to set the password when we want to login in the remote server? In a lot of scene, we already set the password free login. Can you add the password free login to the sshtask?
> > image
> > image
> > image

OK, I will perfect this feature in the next time.  I was changing job recently. Now I  work in a new company and I have times to finish this feature.

**Comment by dailidong** [comment]
image

there are some conflict files need to be sovled, please solve it locally, then I will merge into dev, thx

**Comment by CalvinKirs** [comment]
no update.

**Pull request closed without merging**